Add end of run functionality to CICE

dpsarmie opened the pull request:
## PR checklist
- [X] Short (1 sentence) summary of your PR: 
    This PR will add end of run restart functionality to CICE.
- [X] Developer(s): 
    Daniel Sarmiento
- [x] Suggest PR reviewers from list in the column to the right.
- [X] Please copy the PR test results link or provide a summary of testing completed below.
    This PR was testing in UFS on Hera using the regression test **datm_cdeps_control_gefs**. This was tested using different combinations of restart setting to try and ensure expected functionality.
- How much do the PR code changes differ from the unmodified code? 
    - [x] bit for bit
    - [ ] different at roundoff level
    - [ ] more substantial 
- Does this PR create or have dependencies on Icepack or any other models?
    - [ ] Yes
    - [X] No
- Does this PR update the Icepack submodule?  If so, the Icepack submodule must point to a hash on Icepack's main branch.
    - [ ] Yes
    - [X] No
- Does this PR add any new test cases?
    - [ ] Yes
    - [X] No
- Is the documentation being updated? ("Documentation" includes information on the wiki or in the .rst files from doc/source/, which are used to create the online technical docs at https://readthedocs.org/projects/cice-consortium-cice/. A test build of the technical docs will be performed as part of the PR testing.)
    - [ ] Yes
    - [x] No, does the documentation need to be updated at a later time?
        - [ ] Yes
        - [x] No 
- [X] Please document the changes in detail, including _why_ the changes are made.  This will become part of the PR commit log.

This PR adds the ability for CICE to write restart files at the end of the run even if the end of a run does not coincide with the restart interval. The restart is controlled via the CMEPS configuration option **write_restart_at_endofrun**. If this configuration option is not set or set to _False_, a restart will not be created at the end of the run. Setting this configuration option to _True_ will create a restart file for CMEPS, MOM6, and CICE. The code was written so all three components were tied together by this one configuration variable.

This PR partially resolves ufs-community/ufs-weather-model#2236 and resolves NOAA-EMC/CMEPS#118.

A similar PR is in the MOM6 repository to complete resolve these issues.

dpsarmie committed:
Add end of run restart file functionality

This commit changes the ice_comp_nuopc file to add the ability to
create a restart file at the end of a run. When the stop_alarm is
active, the code will now force a restart file to be written.

dpsarmie committed:
Add CICE write_restart_at_endofrun config option call

This will allow the user to set the write_restart_at_endofrun option
in the ufs.configure.*.IN file to control the creation of the
restart file at the end of the run. This single config option controls
restart files for CMEPS, MOM, and CICE.

DeniseWorthen commented on cicecore/drivers/nuopc/cmeps/ice_comp_nuopc.F90:
```diff
@@ -1131,8 +1133,16 @@ subroutine ModelAdvance(gcomp, rc)
-       force_restart_now = .false.
+    endif
+
+    ! Handle end of run restart
```
I think it is better to do the retrieval of the config variable once (in IntializeAdvertise) and use it set a logical which will enable checking if the stop alarm is ringing. Here, you are retrieving the configuration at each step through ModelAdvance. Look at how it is done in CMEPS, by setting the logical flag in an _init call.

dpsarmie replied:
Yep, that makes sense. Config check has been moved and code has been tested using the different config combinations.

dpsarmie committed:
Move configuration check to IntializeAdvertise

The check for the write_restart_at_endofrun config option was going
to run at each model step. This change will run the check once and
set a logical variable for the subsequent check.

dpsarmie committed:
Turn end of run alarm off 

Turn off stop_alarm check when write at end of run restart is called.

DeniseWorthen commented on cicecore/drivers/nuopc/cmeps/ice_comp_nuopc.F90:
```diff
@@ -1131,8 +1140,15 @@ subroutine ModelAdvance(gcomp, rc)
+    endif
+
+    ! Handle end of run restart
+    if (restart_eor) then
+       call ESMF_ClockGetAlarm(clock, alarmname='alarm_stop', alarm=alarm, rc=rc)
+       if (ChkErr(rc,__LINE__,u_FILE_u)) return
+       if (ESMF_AlarmIsRinging(alarm, rc=rc)) force_restart_now = .true.
+       call ESMF_AlarmRingerOff( alarm, rc=rc )
+       if (ChkErr(rc,__LINE__,u_FILE_u)) return
```
You're not checking the return code for the AlarmIsRinging call. You'll want the same if-then logic as for the restart_alarm check.

Also, how is this generating the correct file name, if you're not getting the timestamp?

dpsarmie replied:
I've added a check for the alarm mimicking the call above it (restart interval section) and tested it for completeness.

Re: Filenames
From what I saw in the code, there's logic to handle when there is no restart filename_spec set here:
https://github.com/dpsarmie/CICE/blob/c9e4679f449e30bb4cc0f22164b4401a8b50f7a6/cicecore/cicedyn/infrastructure/io/io_pio2/ice_restart.F90#L214-L222

So I just relied on that to generate the filename since the end of run restart naming convention won't need to move one timestep forward to generate the correct filename.

![image](https://github.com/NOAA-EMC/CICE/assets/42810219/8230c260-507e-4a48-a365-ce09e532dd39)

dpsarmie committed:
Add check for AlarmIsRinging

Added a check for AlarmIsRinging call.

NickSzapiro-NOAA commented:
@dpsarmie Can you sync with the last commit (9c0cd8d)? I think that's why the GHActions is failing

While named as "write_restart_at_endofrun", is it possible/intended for the restart to be triggered and the models continue to integrate?

dpsarmie committed:
Merge branch 'NOAA-EMC:emc/develop' into feature/endofrun_restart

dpsarmie commented:
> @dpsarmie Can you sync with the last commit ([9c0cd8d](https://github.com/NOAA-EMC/CICE/commit/9c0cd8d3ba5963d01711094a3c81b6358fcb8367))? I think that's why the GHActions is failing
> 
> While named as "write_restart_at_endofrun", is it possible/intended for the restart to be triggered and the models continue to integrate?

@NickSzapiro-NOAA Branch has been updated.

Not 100% sure what you mean. The stop_alarm should only ring on the last timestep and only then will the last restart file be generated. I don't think it would be possible for the models to continue running after the final restart file is written, but the other restart files at the predetermined intervals will still write and allow the model to continue running. (At least that's the intent and how it has been behaving when tested.)

DeniseWorthen commented:
@NickSzapiro-NOAA Dan is correct. We normally write restarts via the restart_n,restart_option settings, so every 12 hours for example. However, if the run continues to a non-interval hour and then stops, there is currently no restart written. CMEPS has an existing 'write-restart-at-end-of-run' configuration option, but it wasn't respected by MOM6 or CICE6.

dpsarmie committed:
Merge branch 'NOAA-EMC:emc/develop' into feature/endofrun_restart

DeniseWorthen commented:
@dpsarmie Could you add this simple change to this PR

```
diff --git a/cicecore/drivers/nuopc/cmeps/cice_wrapper_mod.F90 b/cicecore/drivers/nuopc/cmeps/cice_wrapper_mod.F90
index d0aafbb..4668f03 100644
--- a/cicecore/drivers/nuopc/cmeps/cice_wrapper_mod.F90
+++ b/cicecore/drivers/nuopc/cmeps/cice_wrapper_mod.F90
@@ -72,7 +72,7 @@ subroutine ufs_logfhour(msg,hour)
     real(dbl_kind),    intent(in)    :: hour
     character(len=char_len)          :: filename
     integer(int_kind)                :: nunit
-    write(filename,'(a,i3.3)')'log.ice.f',int(hour)
+    write(filename,'(a,i4.4)')'log.ice.f',int(hour)
     open(newunit=nunit,file=trim(filename))
     write(nunit,'(a)')'completed: cice'
     write(nunit,'(a,f10.3)')'forecast hour:',hour
```

dpsarmie committed:
Update cice_wrapper_mod.F90

Change precision of variable being logged

dpsarmie commented:
@DeniseWorthen Ok added that change.

FernandoAndrade-NOAA commented:
Testing for #[2324](https://github.com/ufs-community/ufs-weather-model/pull/2324) is complete, please continue with the merge process for this PR.

FernandoAndrade-NOAA commented:
Correction - Please do not merge yet, there is still a missing commit on the WM side.

FernandoAndrade-NOAA commented:
Ok, the missing log in 2324 has been committed. We should be all set to continue, sorry for the delay.
